File: tech-dhtml/connect4/models/Player.js
import { Coordinate } from './Board.js';

class Player {

    #color;
    #board;

    constructor(color, board) {
        this.#color = color;
        this.#board = board;
    }
    play(column) {
        this.#board.dropToken(column, this.#color);
    }
    getColor() {
        return this.#color;
    }
    isComplete(column) {
        return this.#board.isComplete(column);
    }
    accept(visitor) { }
}

class HumanPlayer extends Player {

    constructor(color, board) {
        super(color, board);
    }
    accept(visitor) {
        visitor.visitHumanPlayer(this);
    }

}

class RandomPlayer extends Player {

    constructor(color, board) {
        super(color, board);
    }

    accept(visitor) {
        visitor.visitRandomPlayer(this);
    }

    getColumn() {
        let column;
        do {
            column = Math.floor(Math.random() * Coordinate.NUMBER_COLUMNS);
        } while (this.isComplete(column));
        return column;
    }

}

class PlayerVisitor {

    visitHumanPlayer(humanPlayer) {
    }
    visitRandomPlayer(randomPlayer) {
    }

}

export { HumanPlayer, RandomPlayer , PlayerVisitor };

File: tech-java-orientadoObjetos/connecta$/src/connect4/models/Player.java
package connect4.models;

public abstract class Player {
    private Color color;
    private Board board;

    public Player(Color color, Board board) {
        this.color = color;
        this.board = board;
    }

    public void play(int column) {
        this.board.dropToken(column, this.color);
    }

    public Color getColor() {
        return this.color;
    }

    public boolean isComplete(int column) {
        return this.board.isCompleteColumn(column);
    }

    public Board getBoard() {
        return this.board;
    }

    public abstract void accept(PlayerVisitor visitor);
}
